### ISSUE TITLE
Graph break doesn't result in continuation function when break happens in if-statement expression without inline function call

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#111918]
### 🐛 Describe the bug

After https://github.com/pytorch/pytorch/pull/111919 I noticed that I can now trigger a graph break without Dynamo creating a continuation frame. Modify the test function as such:

```
diff --git a/test/dynamo/test_unspec.py b/test/dynamo/test_unspec.py
index b5c3bfd02ca..02b0d48bd8d 100644
--- a/test/dynamo/test_unspec.py
+++ b/test/dynamo/test_unspec.py
@@ -381,7 +381,7 @@ class UnspecTests(torch._dynamo.test_case.TestCase):
         def fn(x):
             x = x + 1
             y = x.item()
-            if test(y):
+            if y > 2:
                 return x * 2
             else:
                 return x * 3
```

(effectively inlining test.) This causes a continuation frame to no longer be created:

<img width="808" alt="image" src="https://github.com/pytorch/pytorch/assets/13564/cca189bb-0c9e-459a-9737-e5ea08682c0f">

If you move the graph break to an inner function call (as the test is written in the PR), this solves the problem. So it's probably some minor issue. I took a quick look though and it wasn't obvious what the issue was.

### Versions

main

cc @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by ezyang** [comment]
cc @jansel

**Comment by jon-chuang** [comment]
I've identified the root cause as:
- inlining failing will cause the inlining to be eagerly traced, thus returning a constant result 
- we know how to graph break on a constant result

We do not retrace the coercion to constant result if we do not inline. Perhaps triggering a restart analysis with require shape[0] to be constant would do the trick

**Comment by ezyang** [comment]
Minimal repro

```
import torch
import torch._dynamo.config

torch._dynamo.config.capture_scalar_outputs = True

@torch.compile(backend="eager", dynamic=True)
def fn(x):
    x = x + 1
    y = x.item()
    if y > 2:
        return x * 2
    else:
        return x * 3

x = torch.tensor([3.0])
fn(x)
```

Run with `TORCH_LOGS=bytecode`

Bad modified bytecode is

```
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode] MODIFIED BYTECODE fn /data/users/ezyang/b/pytorch/ex.py line 6 
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]   6           0 LOAD_GLOBAL              1 (__compiled_fn_0)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]               2 LOAD_FAST                0 (x)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]               4 CALL_FUNCTION            1
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]               6 STORE_FAST               2 (graph_out_0)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]               8 LOAD_FAST                2 (graph_out_0)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              10 LOAD_CONST               4 (0)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              12 BINARY_SUBSCR
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              14 LOAD_FAST                2 (graph_out_0)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              16 LOAD_CONST               1 (1)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              18 BINARY_SUBSCR
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              20 STORE_FAST               1 (y)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              22 STORE_FAST               0 (x)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode] 
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]  10          24 LOAD_FAST                1 (y)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              26 LOAD_CONST               2 (2)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              28 COMPARE_OP               4 (>)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              30 POP_JUMP_IF_FALSE       20 (to 40)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode] 
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]  11          32 LOAD_FAST                0 (x)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              34 LOAD_CONST               2 (2)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              36 BINARY_MULTIPLY
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              38 RETURN_VALUE
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode] 
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]  13     >>   40 LOAD_FAST                0 (x)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              42 LOAD_CONST               3 (3)
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              44 BINARY_MULTIPLY
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]              46 RETURN_VALUE
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode] 
V0322 20:04:22.123000 139625667089408 torch/_dynamo/convert_frame.py:569] [0/0_1] [__bytecode]
```

**Comment by jansel** [comment]
Fixed by #122546
